### PR TITLE
Support Multiple Input Formats for `checkpoint`

### PR DESCRIPTION
**Pull request by Thaurun**
For the `forward` method of `TransformerLayer`, users sometimes define custom inputs without gradients, such as:  

1. In some Visual Transformer models, different `TransformerLayer` may require different `attention_mask` or `rotary_pos_emb`. Here, the input is provided as a `tuple`, allowing the `TransformerLayer` to select the appropriate one.  
2. When users want to use custom attention mechanisms, `attention_mask` might not be a tensor—for example, in `xformers`'s `memory_efficient_attention_forward`.  

In such cases, gradient checkpointing (`recompute`) fails to work. This PR introduces a simple modification to handle these scenarios.  

---

### CODE REVIEW

**Comment by Phlip79** [comment]
**We are changing our review process and marking all open, unlabeled PRs as draft. This change will go in effect starting once #3659 is merged.**

Moving forward, all PRs will be required to start as draft PRs. If you wish to get your PR merged, mark your PR as “Ready for review”. Read more about the new process at [submit.md](https://github.com/Phlip79/Megatron-LM/blob/d3763629bcdd04acca52477af7eb8795090426ff/docs/developer/submit.md).
